File: rust/src/solution876.rs
// Definition for singly-linked list.
#[derive(PartialEq, Eq, Clone, Debug)]
pub struct ListNode {
  pub val: i32,
  pub next: Option<Box<ListNode>>
}

impl ListNode {
  #[inline]
  fn new(val: i32) -> Self {
    ListNode {
      next: None,
      val
    }
  }
}

use crate::common::Solution;

impl Solution {
    pub fn middle_node(head: Option<Box<ListNode>>) -> Option<Box<ListNode>> {
        let cloned_head = head.clone();
        Self::helper(head, cloned_head)
    }

    fn helper(slow: Option<Box<ListNode>>, fast: Option<Box<ListNode>>) -> Option<Box<ListNode>> {
        if fast.is_none() {
            return None;
        }

        let fast_node = fast?;
        let slow_node = slow?;

        if fast_node.next.is_none() {
            return Some(slow_node);
        }

        let fast_next_next = fast_node.next?.next;
        if fast_next_next.is_none() {
            return slow_node.next;
        }

        Self::helper(slow_node.next, fast_next_next)
    }
}

File: java/Solution876.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */
class Solution {
    public ListNode middleNode(ListNode head) {
        // slow and fast

        /**
            f         f         f
            s    s    s
            1 -> 2 -> 3 -> 4 -> 5

            f         f         f          f
            s    s    s    s
            1 -> 2 -> 3 -> 4 -> 5 -> 6 -> null

            f          f
            s    s
            1 -> 2 -> null
        */
        ListNode slow = head;
        ListNode fast = head;

        while (fast != null && fast.next != null) {
            slow = slow.next;
            fast = fast.next.next;
        }

        return slow;
    }
}